docs/architecture-decisions/2023-06-06_frontend-templates.md

```diff
@@ -0,0 +1,133 @@
+
+## The proposal
+
+The proposal consists of 2 main parts. On a high level, the ideas revolve around giving developers a choice on how they want to couple AlgoKit backend and frontend templates if they like to use custom templates. The other is focused on providing an official full-stack template that bundles both backend and frontend templates together, while also serving as an example for template builders on how to efficiently couple their backend and frontend templates.
+
+The addendums to the proposal also explores orthogonal ideas that can further improve the CLI tooling itself by providing a way for any existing non web-3 frontend project be converted into a full-stack dApp with minimal efforts. Lastly, it expands on improving incentives for developers to build and maintain their own templates.
+
+---
+
+### Part 1. Independent frontend templates
+
+> TLDR: Independent frontend templates will be created to provide developers with a highly customizable dApp starter project, built on AlgoKit's principles of modularity, maintainability, and flexibility. The templates will also serve as a reference for template builders.
+
+The following aims to provide a seamless onramp for developers to get started with highly customizable dApp starter projects. The idea is to create a set of separate official frontend template repositories to serve as:
+a) A reference for template builders on how to create standalone frontend templates that can be then further coupled with any backend template.
+b) Expand on AlgoKit principles of modularity, maintainability and flexibility by giving developers a choice of preffered technological stack.
+
+The official standalone frontend templates can be build by reusing already established best practices and templates from official backend repositores and by continuing reliance on `copier` for template automation. Another important consideration to keep in mind is that with the introduction of frontend templates we need to establish a clear separation of concerns between the backend and frontend templates to ensure modularity.
+
+![Diagram 1](assets/2023-06-06_frontend-templates/modular_templates.jpg)
+
+As demonstrated on the diagram above, the only glue connecting the backend and frontend is the generated typed client. Neither backend or frontend templates should be concerned with the other but instead provide modular interfaces that clearly indicate to developers on how to integrate the two. From a perspective of a backend template, the typed client can be seen as a `public` folder in modern web frameworks, a static asset that can be reused by any frontend template. Frontend templates on the other hand are mostly standard web projects with an additional layer of utilities that optionally allow them to be integrated with typed clients produced by backend templates.
```

Comment:
> the typed client can be seen as a `public` folder in modern web frameworks

I think I understand the point you are making, but given it will be TypeScript code it can't actually be in the public folder of most web frameworks since that is packaged as-is rather than compiled into JavaScript.

Reply:
I removed the sentence with public folder to avoid confusion, this wasn't implying that we store them in public folder in literal sense but more of an analogy. As with most web frameworks you can expect existence of public folder with static assets - when dealing with algokit backend templates you should expect existence of a special folder that hosts typed clients